### PR TITLE
Fix vllm_importance_sampling_correction sequence-level aggregation: sum → mean

### LINKED ISSUE #5814: Vllm_importance_sampling_correction sequence-level mode aggregates per-token log-ratios with sum instead of mean
### Reproduction

  In `trl/trainer/grpo_trainer.py`, the `vllm_importance_sampling_correction` block aggregates per-token log-probability differences with **sum** instead of **mean** when
  `vllm_importance_sampling_mode` is `sequence_mask` or `sequence_truncate`. The same file uses **mean** for the policy IS at `importance_sampling_level="sequence"` (line ~2501).
  This inconsistency produces an IS ratio whose variance scales with completion length, and combined with `clamp(max=cap)` it deflates the effective gradient even when per-token
  logprobs from vLLM and the trainer agree.

Bug location: 
  ```python
  # trl/trainer/grpo_trainer.py, line ~2070 (main as of 2026-05-22)
  sequence_level_is = self.vllm_importance_sampling_mode in ["sequence_mask", "sequence_truncate"]
  if sequence_level_is:
      per_sequence_logps_diff = per_token_logps_diff.sum(dim=-1, keepdim=True)   # ← bug
      logps_diff = per_sequence_logps_diff
  ```

## Expected vs Actual

  - **Expected**: with `num_iterations=1` (vLLM-policy = trainer-policy at gradient time) and matching per-token logprobs, the IS ratio mean should be ≈ 1.0 with tight spread.
  - **Actual**: IS ratio mean lands at **0.4–0.7** with wide spread (0.0 to cap), even when `sampling/sampling_logp_difference/mean` is near zero.

This is a numerical bug in the aggregation, so a torch-only synthetic reproducer is exact:

  ```python
  import torch

  torch.manual_seed(0)
  B, T, cap = 16, 200, 3.0  # 16 sequences, 200 completion tokens, default cap
  mask = torch.ones(B, T)
  # Realistic numerical noise between vLLM and trainer recompute: mean=0, std≈0.1
  per_token_logps_diff = torch.randn(B, T) * 0.1 * mask

  # Current TRL behavior (trainer/grpo_trainer.py:2070)
  seq_diff_sum  = per_token_logps_diff.sum(dim=-1, keepdim=True)
  ratio_sum     = torch.exp(seq_diff_sum).clamp(max=cap)

  # Proposed fix (mean — matches the existing policy-IS aggregation at line ~2501)
  seq_len       = mask.sum(dim=-1, keepdim=True).clamp(min=1.0)
  seq_diff_mean = per_token_logps_diff.sum(dim=-1, keepdim=True) / seq_len
  ratio_mean    = torch.exp(seq_diff_mean).clamp(max=cap)

  print(f"sum-based  IS ratio: mean={ratio_sum.mean():.3f}  std={ratio_sum.std():.3f}  "
        f"min={ratio_sum.min():.3f}  max={ratio_sum.max():.3f}")
  print(f"mean-based IS ratio: mean={ratio_mean.mean():.3f}  std={ratio_mean.std():.3f}  "
        f"min={ratio_mean.min():.3f}  max={ratio_mean.max():.3f}")
  ```

  Output:

  ```
  sum-based  IS ratio: mean=1.286  std=1.166  min=0.121  max=3.000
  mean-based IS ratio: mean=0.999  std=0.008  min=0.990  max=1.014
  ```

### System Info

  - **TRL version:** 1.4.0 
  - **Transformers:** 5.8.0
  - **PyTorch:** 2.10.0 + CUDA 12.8
  - **vLLM:** 0.18.0 

### Checklist

- [x] I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))
- [x] I have included my system information
- [x] Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any traceback provided is complete

### PR DESCRIPTION
**Pull request by ionut-anghelina**
# What does this PR do?

Fixes #5814.

Sequence-level vLLM importance sampling correction was aggregating per-token log-ratios with `sum` instead of `mean`. This made the IS ratio's variance scale with completion length and, combined with `clamp(max=cap)`, systematically deflated the ratio toward zero even when per-token logprobs from vLLM and the trainer agreed.

The fix divides by the mask sum, matching:

- The GSPO paper formula (Zheng et al., 2025), which has a `1/|y|` factor.
- TRL's own policy-IS aggregation at `importance_sampling_level="sequence"` (`grpo_trainer.py` line ~2501), which already uses `(log_ratio * mask).sum(-1) / mask.sum(-1).clamp(min=1.0)`.

Empirically on a real training run, the IS ratio mean went from **0.569** (with `sum`) to **0.993** (with `mean`) — same checkpoint, same seed, same data, bit-identical per-token logp diffs. Synthetic reproducer in #5814 shows the same effect deterministically in <1s.

The change is one line.

## Before submitting

- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [x] Was this discussed/approved via a GitHub issue? Please add a link to it if that's the case. — #5814
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

> The existing test `test_train_vllm_importance_sampling_correction` is marked `@pytest.mark.skip(reason="We should add a mock for the vLLM server.")`, so I have not added a runtime test. Happy to add one if the maintainers point me to the preferred mocking approach.

## AI writing disclosure

- [ ] No AI usage: the PR was written entirely by a human.
- [x] AI-assisted: some parts were suggested or improved by AI, but the PR was written and reviewed by a human.
- [ ] AI-generated: the PR was mostly or fully generated by an AI tool.

## Who can review?

Anyone familiar with GSPO / GRPO importance sampling.

---

> [!NOTE]
> **Medium Risk**
> Adjusts the vLLM importance-sampling correction used during GRPO training, which can materially change loss scaling and training dynamics when `vllm_importance_sampling_mode` is sequence-level. Change is localized but affects core RL training behavior.
> 
> **Overview**
> Fixes sequence-level `vllm_importance_sampling_correction` aggregation in `GRPOTrainer` by normalizing the summed per-token log-prob diffs by the masked completion length (sum → mean). This makes the vLLM importance-sampling ratio in `sequence_*` modes independent of completion length before applying the configured cap/masking behavior.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 6c0afd3d2fa631f6a1035275ed6147af9b0827c3. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by qgallouedec** [comment]
see https://github.com/huggingface/trl/issues/5814#issuecomment-4520944425
